daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1286,37 +1299,51 @@ private[lf] object SBuiltin {
-              onLedger.ptx = onLedger.ptx.copy(keys = onLedger.ptx.keys.updated(gkey, keyMapping))
-              operation.handleKnownInputKey(machine, gkey, keyMapping)
+            case Some(optCid) =>
+              val optActiveCid = optCid match {
+                case KeyActive(cid) if onLedger.ptx.consumedBy.contains(cid) => KeyInactive
+                case _ => optCid
+              }
+              onLedger.ptx = onLedger.ptx.copy(keys = onLedger.ptx.keys.updated(gkey, optActiveCid))
+              operation.handleKnownInputKey(machine, gkey, optActiveCid)
             case None =>
               // if we cannot find it here, send help, and make sure to update [[PartialTransaction.key]] after
               // that.
               throw SpeedyHungry(
                 SResultNeedKey(
                   GlobalKeyWithMaintainers(gkey, keyWithMaintainers.maintainers),
                   onLedger.committers,
-                  {
-                    case SKeyLookupResult.Found(cid) =>
-                      val keyMapping = PartialTransaction.KeyActive(cid)
-                      onLedger.ptx = onLedger.ptx.copy(
-                        keys = onLedger.ptx.keys.updated(gkey, keyMapping),
-                        globalKeyInputs = onLedger.ptx.globalKeyInputs.updated(gkey, keyMapping),
-                      )
-                      // We have to check that the discriminator of cid does not conflict with a local ones
-                      // however we cannot raise an exception in case of failure here.
-                      // We delegate to SEImportValue the task to check cid.
-                      operation.handleInputKeyFound(machine, cid)
-                      true
-                    case SKeyLookupResult.NotFound =>
-                      val keyMapping = PartialTransaction.KeyInactive
-                      onLedger.ptx = onLedger.ptx.copy(
-                        keys = onLedger.ptx.keys.updated(gkey, keyMapping),
-                        globalKeyInputs = onLedger.ptx.globalKeyInputs.updated(gkey, keyMapping),
-                      )
-                      operation.handleInputKeyNotFound(machine)
-                    case SKeyLookupResult.NotVisible =>
-                      machine.tryHandleSubmitMustFail()
+                  result => {
+                    cacheGlobalLookup(onLedger, gkey, result)
+                    // We need to check if the contract was consumed since we only
+                    // modify keys if the archive was for a key
+                    // already brought into scope.
+                    val activeResult: SKeyLookupResult = result match {
+                      case SKeyLookupResult.Found(cid) =>
+                        if (onLedger.ptx.consumedBy.contains(cid))
+                          SKeyLookupResult.NotFound
+                        else result
+                      case SKeyLookupResult.NotFound | SKeyLookupResult.NotVisible =>
+                        result
+                    }
```

Comment:
_(cosmetic)_

Why not 
```suggestion
                    val activeResult = result match {
                      case SKeyLookupResult.Found(cid) if onLedger.ptx.consumedBy.contains(cid) =>
                          SKeyLookupResult.NotFound
                      case SKeyLookupResult.Found(_) | SKeyLookupResult.NotFound | SKeyLookupResult.NotVisible =>
                        result
                    }
```

or simply 

```suggestion
                    val activeResult = result match {
                      case SKeyLookupResult.Found(cid) if onLedger.ptx.consumedBy.contains(cid) =>
                          SKeyLookupResult.NotFound
                      case _ =>
                        result
                    }
```